segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/GarbageCollector.java

```diff
@@ -132,240 +160,339 @@ public GarbageCollector(int containerId, ChunkStorage chunkStorage,
+    public CompletableFuture<Void> initialize(AbstractTaskQueue<TaskInfo> taskQueue) {
+        // Temp change to be removed after next PR (part 2 of 3)
+        if (null == taskQueue) {
+            return CompletableFuture.completedFuture(null);
+        }
+
+        this.taskQueue = Preconditions.checkNotNull(taskQueue, "taskQueue");
+        return taskQueue.addQueue(this.taskQueueName, false)
+                .thenComposeAsync(v -> taskQueue.addQueue(this.failedQueueName, true), storageExecutor);
     }
 
     /**
-     * Gets a value indicating how much to wait for the service to shut down, before failing it.
+     * Adds given chunks to list of garbage chunks.
      *
-     * @return The Duration.
+     * @param chunksToDelete List of chunks to delete.
      */
-    @Override
-    protected Duration getShutdownTimeout() {
-        return SHUTDOWN_TIMEOUT;
+    CompletableFuture<Void> addChunksToGarbage(long transactionId, Collection<String> chunksToDelete) {
+        if (null != taskQueue) {
+            val currentTime = currentTimeSupplier.get();
+            val futures = new ArrayList<CompletableFuture<Void>>();
+            chunksToDelete.forEach(chunkToDelete -> futures.add(addChunkToGarbage(transactionId, chunkToDelete, currentTime + config.getGarbageCollectionDelay().toMillis(), 0)));
+            return Futures.allOf(futures);
+        }
+        return CompletableFuture.completedFuture(null);
     }
 
     /**
-     * Main execution of the Service. When this Future completes, the service auto-shuts down.
+     * Adds given chunk to list of garbage chunks.
      *
-     * @return A CompletableFuture that, when completed, indicates the service is terminated. If the Future completed
-     * exceptionally, the Service will shut down with failure, otherwise it will terminate normally.
+     * @param chunkToDelete Name of the chunk to delete.
+     * @param startTime Start time.
+     * @param attempts Number of attempts to delete this chunk so far.
      */
-    @Override
-    protected CompletableFuture<Void> doRun() {
-        loopFuture = Futures.loop(
-                this::canRun,
-                () -> delaySupplier.get()
-                        .thenComposeAsync(v -> deleteGarbage(true, config.getGarbageCollectionMaxConcurrency()), executor)
-                        .handleAsync((v, ex) -> {
-                            if (null != ex) {
-                                log.error("{}: Error during doRun.", traceObjectId, ex);
-                            }
-                            return null;
-                        }, executor),
-                executor);
-        return loopFuture;
+    CompletableFuture<Void>  addChunkToGarbage(long transactionId, String chunkToDelete, long startTime, int attempts) {
+        if (null != taskQueue) {
+            return taskQueue.addTask(taskQueueName, new TaskInfo(chunkToDelete, startTime, attempts, TaskInfo.DELETE_CHUNK, transactionId))
+                    .thenRunAsync(() -> {
```

Comment:
Does this need to be `Async`? Looks like a very small method.

Same below.

Reply:
yes it needs to Async as the underlying implementation of taskQueue is using EventProcessor async API